### PR TITLE
MGMT-4583: Refactor ocs code

### PR DESCRIPTION
**Pull request by rewantsoni**
Signed-off-by: Rewant Soni <resoni@redhat.com>

### CODE REVIEW

File: internal/operators/ocs/consts.go [hunk 1]
```diff
@@ -1,7 +1,12 @@
...
 
 const (
-	ssdDrive    string = "SSD"
-	hddDrive    string = "HDD"
-	minDiskSize int64  = 5 //5GB is the min disk size for OCS
+	ssdDrive              string = "SSD"
+	hddDrive              string = "HDD"
+	minDiskSize           int64  = 5 //5GB is the min disk size for OCS
+	compactMode           string = "Compact Mode"
+	compact               string = "Compact"
```
Comment (priyanka19-98) [inline internal/operators/ocs/consts.go:8]: Any reason to have `compact` twice?

File: internal/operators/ocs/consts.go [hunk 2]
```diff
@@ -1,7 +1,12 @@
...
+	minimal               string = "Minimal"
+	standard              string = "Standard"
+	minimalDeploymentMode        = "Minimal Deployment Mode"
```
Comment (priyanka19-98) [inline internal/operators/ocs/consts.go:11]: similarly here

File: internal/operators/ocs/consts.go (hunk 1 above)
Comment (rewantsoni) [reply internal/operators/ocs/consts.go]: fixed

File: internal/cluster/validations/validations.go [hunk 3]
```diff
@@ -40,7 +40,6 @@ const (
 
 // PullSecretValidator is used run validations on a provided pull secret
 // it verifies the format of the pull secrete and access to required image registries
-// go:generate mockgen -source=validations.go -package=validations -destination=mock_validations.go
```
Comment (nbalacha) [inline internal/cluster/validations/validations.go:43]: Why has this been removed?

Comment (rewantsoni) [reply internal/cluster/validations/validations.go:43]: we don't need mock_validation.go now as it's not called anywhere, so I removed it

Comment (priyanka19-98) [reply internal/cluster/validations/validations.go:43]: @rewantsoni , I guess this interface refers to https://github.com/openshift/assisted-service/blob/master/internal/cluster/validations/mock_validations.go and has nothing to do with OCS validations. Please restore the changes again

Comment (rewantsoni) [reply internal/cluster/validations/validations.go:43]: fixed

File: internal/operators/ocs/validations.go [hunk 4]
```diff
@@ -44,35 +27,31 @@ type Config struct {
...
-	var inventoryMissing bool
 	hosts := cluster.Hosts
 
-	if int64(len(hosts)) < o.OCSRequiredHosts {
+	//Check if number of hosts is less than required for compact mode
+	if int64(len(hosts)) < o.config.OCSRequiredHosts {
 		status = "Insufficient hosts to deploy OCS. A minimum of 3 hosts is required to deploy OCS. "
 		o.log.Info("Validate Requirements status ", status)
 		return api.Failure, status
 	}
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:40]: Can we move this part to a function called `checkOCSRequiredHostsLength` ?

File: internal/operators/ocs/validations.go [hunk 5]
```diff
@@ -44,35 +27,31 @@ type Config struct {
...
-				return api.Pending, "Missing Inventory in some of the hosts"
+				return res, err.Error()
 			}
 		}
-
-	} else if len(hosts) < o.OCSMasterWorkerHosts { // not supporting OCS installation for 2 Workers and 3 Masters
+	} else if len(hosts) < o.config.OCSMasterWorkerHosts { // not supporting OCS installation for 2 Workers and 3 Masters
 		status = "Not supporting OCS Installation for 3 Masters and 2 Workers"
 		o.log.Info("Validate Requirements status ", status)
 		return api.Failure, status
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:57]: Similarly this to `checkOCSRequiredHostsLength`

File: internal/operators/ocs/validations.go [hunk 6]
```diff
@@ -142,17 +128,17 @@ func (o *ocsValidator) nodeResources(host *models.Host, cpuCount *int64, totalRA
...
 			*insufficientHosts = append(*insufficientHosts, status)
 			o.log.Info(status)
 		} else {
-			*cpuCount += (inventory.CPU.Count - requiredDiskCPU)                     // cpus excluding the cpus required for disks
-			*totalRAM += (inventory.Memory.UsableBytes - gbToBytes(requiredDiskRAM)) // ram excluding the ram required for disks
-			*diskCount += (int64)(disks - 1)                                         // not counting the boot disk
+			*cpuCount += inventory.CPU.Count - requiredDiskCPU                     // cpus excluding the cpus required for disks
+			*totalRAM += inventory.Memory.UsableBytes - gbToBytes(requiredDiskRAM) // ram excluding the ram required for disks
+			*diskCount += (int64)(disks - 1)                                       // not counting the boot disk
 			*hostsWithDisks++
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:142]: And this part to different function?

File: internal/operators/ocs/validations.go [hunk 7]
```diff
@@ -128,7 +103,18 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
 	return api.Failure, status
 }
 
-func (o *ocsValidator) nodeResources(host *models.Host, cpuCount *int64, totalRAM *int64, diskCount *int64, hostsWithDisks *int64, insufficientHosts *[]string) (bool, error) {
+func checkInventoryAndNodeRequirements(o *operator, host *models.Host, cpuCount *int64, totalRAM *int64, diskCount *int64, hostsWithDisks *int64, insufficientHosts *[]string) (api.ValidationStatus, error) {
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:106]: Instead of passing `o *operator` , include that in function definition, so that you don't need to pass explicitly

**Review by nbalacha — CHANGES_REQUESTED** [review]
This is the first set of review comments.
I am not done with reviewing validations.go and will send those review comments in a while.

File: internal/operators/ocs/validations.go (hunk 6 above)
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:142]: done

File: internal/operators/ocs/validations.go (hunk 7 above)
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:106]: done

File: internal/operators/ocs/manifest.go [hunk 8]
```diff
@@ -29,22 +29,22 @@ func OCSStorageCluster(StorageClusterManifest string, ocsDiskCounts int64) ([]by
 
 }
 
-func Manifests(ocsMinDeploy bool, ocsDiskCounts int64, totalHosts int) (map[string][]byte, error) {
+func manifests(ocsMinDeploy bool, ocsDiskCounts int64, totalHosts int) (map[string][]byte, error) {
```
[the remark this answers is not in the document]
Comment (rewantsoni) [reply internal/operators/ocs/manifest.go:32]: done

File: internal/operators/ocs/validations.go [hunk 9]
```diff
@@ -44,35 +27,31 @@ type Config struct {
...
 	var status string
-	var err error
-	var inventoryMissing bool
 	hosts := cluster.Hosts
 
-	if int64(len(hosts)) < o.OCSRequiredHosts {
+	//Check if number of hosts is less than required for compact mode
+	if int64(len(hosts)) < o.config.OCSRequiredHosts {
 		status = "Insufficient hosts to deploy OCS. A minimum of 3 hosts is required to deploy OCS. "
 		o.log.Info("Validate Requirements status ", status)
```
[the remark this answers is not in the document]
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:38]: It would translate to `Validate Requirements status Insufficient hosts to deploy OCS. A minimum of 3 hosts is required to deploy OCS. ` it does contain OCS, should we change it to a better message?

File: internal/operators/ocs/validations.go [hunk 10]
```diff
@@ -44,35 +27,31 @@ type Config struct {
...
 }
 
-// ValidateOCSRequirements is used to validate min requirements of OCS
-func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.ValidationStatus, string) {
+// validateRequirements is used to validate min requirements of OCS
+func (o *operator) validateRequirements(cluster *models.Cluster) (api.ValidationStatus, string) {
 	var status string
-	var err error
-	var inventoryMissing bool
 	hosts := cluster.Hosts
```
[the remark this answers is not in the document]
Comment (rewantsoni) [reply internal/operators/ocs/validations.go:34]: done

File: internal/operators/ocs/validations.go [hunk 11]
```diff
@@ -128,40 +123,55 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
...
 	} else {
-		*cpuCount += inventory.CPU.Count
-		*totalRAM += inventory.Memory.UsableBytes
+		*r.cpuCount += inventory.CPU.Count
+		*r.totalRAM += inventory.Memory.UsableBytes
 	}
 	return false, nil
 }
 
+func (o *operator) countResources(inventory *models.Inventory, r resources, disks int64) {
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:159]: rename to `countResourcesExcludingDiskResources` and add a comment that `OCS requires 2 cpus, 5GB ram` per disk

File: internal/operators/ocs/validations.go [hunk 12]
```diff
@@ -182,64 +192,69 @@ func getValidDiskCount(disks []*models.Disk) int {
...
 		status = status + fmt.Sprint(TotalCPUs, " CPUs, excluding disk CPU resources ")
 	}
 	if ram < gbToBytes(TotalRAMGB) {
 		status = status + fmt.Sprint(TotalRAMGB, " RAM GB, excluding disk RAM resources ")
 	}
-	if disk < o.OCSRequiredDisk {
-		status = status + fmt.Sprint(o.OCSRequiredDisk, " Disks of minimum ", minDiskSize, "GB is required.")
+	if disk < o.config.OCSRequiredDisk {
+		status = status + fmt.Sprint(o.config.OCSRequiredDisk, " Disks of minimum ", minDiskSize, "GB is required.")
 	}
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:255]: remove the `.` after `GB is required` and add a space (` `) instead

File: internal/operators/ocs/validations.go [hunk 13]
```diff
@@ -128,40 +123,55 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
 	return api.Failure, status
 }
 
-func (o *ocsValidator) nodeResources(host *models.Host, cpuCount *int64, totalRAM *int64, diskCount *int64, hostsWithDisks *int64, insufficientHosts *[]string) (bool, error) {
+func (o *operator) checkInventoryAndNodeRequirements(r resources) (api.ValidationStatus, error) {
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:126]: rename to `checkInventoryStatusAndNodeRequirements`

File: internal/operators/ocs/validations.go [hunk 14]
```diff
@@ -128,40 +123,55 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
...
+	if err != nil {
+		o.log.Error("Error occurred while calculating Node requirements ", err)
+		return api.Failure, errors.New("Error occurred while calculating Node requirements ")
+	} else if inventoryMissing {
+		return api.Pending, errors.New("Missing Inventory in some of the hosts ")
+	}
+	return "", nil
+}
+
+func (o *operator) nodeResources(r resources) (bool, error) {
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:137]: Rename to `countNodeResources`

File: internal/operators/ocs/validations.go [hunk 15]
```diff
@@ -44,35 +27,50 @@ type Config struct {
...
+		diskCount:         &diskCount,
+		hostsWithDisks:    &hostsWithDisks,
+		insufficientHosts: &insufficientHosts,
+	}
+
+	if int64(numAvailableHosts) == o.config.OCSRequiredHosts { // for only 3 hosts, we need to install OCS in compact mode
 		for _, host := range hosts {
-			inventoryMissing, err = o.nodeResources(host, &cpuCount, &totalRAM, &diskCount, &hostsWithDisks, &insufficientHosts)
+			r.host = host
+			res, err := o.checkInventoryStatus(r)
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:68]: As you are modifying resources in other functions, can you pass it as a pointer? Not sure if the values are retained when you exit from the function.

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:68]: Also rename `res` to `status`

Comment (rewantsoni) [reply internal/operators/ocs/validations.go:68]: Done

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:68]: You can pass `r` as it is, I did not notice the `cpu, ram` etc were already passed as a pointer. You can ignore my previous comment on passing it as a pointer.

File: internal/operators/ocs/validations.go [hunk 16]
```diff
@@ -128,40 +123,56 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
...
 
+func (o *operator) countResourcesExcludingDiskResources(inventory *models.Inventory, r resources, disks int64) {
+	// OCS requires 2 cpus, 5GB ram per disk
+	requiredDiskCPU := (disks - 1) * o.config.OCSRequiredDiskCPUCount
+	requiredDiskRAM := (disks - 1) * o.config.OCSRequiredDiskRAMGB
+
+	if inventory.CPU.Count < requiredDiskCPU || inventory.Memory.UsableBytes < gbToBytes(requiredDiskRAM) {
+		status := fmt.Sprint("Insufficient resources on host with host ID ", *r.host.ID, " to deploy OCS. The hosts has ", disks, " disks that require ", requiredDiskCPU, " CPUs, ", requiredDiskRAM, " RAMGB")
+		*r.insufficientHosts = append(*r.insufficientHosts, status)
+		o.log.Info(status)
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:167]: Unrelated to this patch but shouldn't this error out here?

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:167]: We wanted to show all the hosts that did not have sufficient disk resources to the user

File: internal/operators/ocs/validations.go [hunk 17]
```diff
@@ -128,40 +123,56 @@ func (o *ocsValidator) ValidateRequirements(cluster *models.Cluster) (api.Valida
 	return api.Failure, status
 }
 
-func (o *ocsValidator) nodeResources(host *models.Host, cpuCount *int64, totalRAM *int64, diskCount *int64, hostsWithDisks *int64, insufficientHosts *[]string) (bool, error) {
+func (o *operator) checkInventoryStatus(r resources) (api.ValidationStatus, error) {
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:126]: This function is not required - its logic can be combined with that of the other node resource checks

File: internal/operators/ocs/validations.go [hunk 18]
```diff
@@ -182,64 +193,69 @@ func getValidDiskCount(disks []*models.Disk) int {
...
 	}
 
 	status = "OCS Requirements for Standard Deployment are satisfied"
-	o.OCSDeploymentType = "Standard"
+	o.config.OCSDeploymentType = standard
 	return true, status
 }
 
-func (o *ocsValidator) setStatusInsufficientResources(cpu int64, ram int64, disk int64, hostsWithDisks int64, mode string) string {
+func (o *operator) workerNodesRequirements(cpu int64, ram int64, disk int64, hostsWithDisks int64, TotalCPUs int64, TotalRAM int64) bool {
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:233]: Misleading function name - the function is called for compact mode as well which means the nodes are masters. 
Other changes:
Change the fn signature to take the resources struct as an argument
Pass in the deployment mode instead of TotalRAM and TotalCPUs. Let the function determine that based on the mode. That way we can extend the function to compare more params in future without changing the signature.

It can be written like this:

func requirementSatisfied(o *operator, nodeInfo *resourceInfo, mode string) bool {
	var totalCPUs int64
	var totalRAM int64

	if mode == compact {
		totalCPUs = o.config.OCSRequiredCompactModeCPUCount
		totalRAM = o.config.OCSRequiredCompactModeRAMGB
	} else if mode == minimal {
		totalCPUs = o.config.OCSMinimumCPUCount
		totalRAM = o.config.OCSMinimumRAMGB
	} else if mode == standard {
		totalCPUs = o.config.OCSRequiredCPUCount
		totalRAM = o.config.OCSRequiredRAMGB
	}
	return nodeInfo.cpuCount >= totalCPUs && nodeInfo.ram >= gbToBytes(totalRAM) && nodeInfo.numDisks >= o.config.OCSRequiredDisk && nodeInfo.hostsWithDisks >= o.config.OCSRequiredHosts
}


And called by validate()  like this:


	if numAvailableHosts == o.config.OCSRequiredHosts { // for 3 hosts
		if !requirementSatisfied(o, nodeInfo, "compact") { // check for master nodes requirements
			status = o.setStatusInsufficientResources(nodeInfo, compact)
			return false, status
		}
		o.config.OCSDeploymentType = compact
		status = "OCS Requirements for Compact Mode are satisfied"
		return true, status
	}

	if !requirementSatisfied(o, nodeInfo, standard) { // check for worker nodes requirement
		if !requirementSatisfied(o, nodeInfo, minimal) { // check for worker nodes requirements
			status = o.setStatusInsufficientResources(nodeInfo, minimal)
			return false, status
		}
		status = "Requirements for OCS Minimal Deployment are satisfied"
		o.config.OCSDeploymentType = minimal
		o.config.OCSMinimalDeployment = true
		return true, status
	}
	status = "OCS Requirements for Standard Deployment are satisfied"
	o.config.OCSDeploymentType = standard
	return true, status
}

File: internal/operators/ocs/validations.go [hunk 19]
```diff
@@ -182,64 +193,69 @@ func getValidDiskCount(disks []*models.Disk) int {
...
+	return cpu >= TotalCPUs && ram >= gbToBytes(TotalRAM) && disk >= o.config.OCSRequiredDisk && hostsWithDisks >= o.config.OCSRequiredHosts
+}
+
+func (o *operator) setStatusInsufficientResources(cpu int64, ram int64, disk int64, hostsWithDisks int64, mode string) string {
 	var TotalCPUs int64
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:238]: Please change all local variable names to start with small case

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: internal/operators/ocs/ocs_operator.go [hunk 20]
```diff
@@ -27,99 +26,98 @@ var Operator = models.MonitoredOperator{
...
 	return api.ValidationResult{Status: api.Success, ValidationId: o.GetHostValidationID(), Reasons: []string{}}, nil
 }
 
 // GenerateManifests generates manifests for the operator
-func (o *ocsOperator) GenerateManifests(cluster *common.Cluster) (map[string][]byte, error) {
-	o.log.Info("No. of OCS eligible disks are ", o.ocsValidatorConfig.OCSDisksAvailable)
-	return Manifests(o.ocsValidatorConfig.OCSMinimalDeployment, o.ocsValidatorConfig.OCSDisksAvailable, len(cluster.Cluster.Hosts))
+func (o *operator) GenerateManifests(cluster *common.Cluster) (map[string][]byte, error) {
+	o.log.Info("No. of OCS eligible disks are ", o.config.OCSDisksAvailable)
+	//return manifests(o.config.OCSMinimalDeployment, o.config.OCSDisksAvailable, len(cluster.Cluster.Hosts))
```
[the remark this answers is not in the document]
Comment (nbalacha) [reply internal/operators/ocs/ocs_operator.go:81]: Please remove this comment

File: internal/operators/ocs/validation_test.go [hunk 21]
```diff
@@ -447,12 +447,12 @@ var _ = Describe("Ocs Operator use-cases", func() {
 				clust.IsDNSDomainDefined:                  {status: clust.ValidationSuccess, messagePattern: "The base domain is defined"},
 				clust.IsPullSecretSet:                     {status: clust.ValidationSuccess, messagePattern: "The pull secret is set"},
 				clust.SufficientMastersCount:              {status: clust.ValidationSuccess, messagePattern: "The cluster has a sufficient number of master candidates."},
-				clust.IsOcsRequirementsSatisfied:          {status: clust.ValidationFailure, messagePattern: "Total disks on the cluster must be a multiple of 3 of size >= 5 GB"},
+				clust.IsOcsRequirementsSatisfied:          {status: clust.ValidationFailure, messagePattern: "Total disks on the cluster must be a multiple of 3 of size >= 5 GB >= 5 GB"},
```
Comment (nbalacha) [inline internal/operators/ocs/validation_test.go:450]: Please remove the extra ">= 5GB" from the message

File: internal/operators/ocs/validation_test.go [hunk 22]
```diff
@@ -462,7 +462,7 @@ var _ = Describe("Ocs Operator use-cases", func() {
 			pullSecretSet:      true,
 			hosts: []models.Host{
 				{ID: &hid1, Status: swag.String(models.HostStatusKnown), Inventory: ocsInventoryWithDisks(6, 64000000000), Role: models.HostRoleMaster},
-				{ID: &hid2, Status: swag.String(models.HostStatusKnown), Inventory: ocsInventoryWithDisks(7, 64000000000), Role: models.HostRoleMaster},
+				{ID: &hid2, Status: swag.String(models.HostStatusKnown), Inventory: ocsInventoryWithDisks(5, 64000000000), Role: models.HostRoleMaster},
```
Comment (nbalacha) [inline internal/operators/ocs/validation_test.go:465]: Why is this change required?

File: internal/operators/ocs/validation_test.go [hunk 23]
```diff
@@ -596,7 +596,7 @@ var _ = Describe("Ocs Operator use-cases", func() {
 				clust.IsDNSDomainDefined:                  {status: clust.ValidationSuccess, messagePattern: "The base domain is defined"},
 				clust.IsPullSecretSet:                     {status: clust.ValidationSuccess, messagePattern: "The pull secret is set"},
 				clust.SufficientMastersCount:              {status: clust.ValidationSuccess, messagePattern: "The cluster has a sufficient number of master candidates."},
-				clust.IsOcsRequirementsSatisfied:          {status: clust.ValidationFailure, messagePattern: "Total disks on the cluster must be a multiple of 3 of size >= 5 GB"},
+				clust.IsOcsRequirementsSatisfied:          {status: clust.ValidationFailure, messagePattern: "Total disks on the cluster must be a multiple of 3 of size >= 5 GB >= 5 GB"},
```
Comment (nbalacha) [inline internal/operators/ocs/validation_test.go:599]: Please remove the extra ">= 5GB"

File: internal/operators/ocs/validation_test.go [hunk 24]
```diff
@@ -785,7 +785,7 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
```
Comment (nbalacha) [inline internal/operators/ocs/validation_test.go:788]: Please remove the extra ">= 5GB"

File: internal/operators/ocs/validation_test.go [hunk 25]
```diff
@@ -818,7 +818,7 @@ var _ = Describe("Ocs Operator use-cases", func() {
...
```
Comment (nbalacha) [inline internal/operators/ocs/validation_test.go:821]: Same as above

File: internal/operators/ocs/validations.go [hunk 26]
```diff
@@ -2,247 +2,251 @@ package ocs
...
+
+	if canDeployOCS {
+		// this will be used to set count of StorageDevices in StorageCluster manifest
+		o.config.OCSDisksAvailable = nodesResourceInfo.numDisks
+		return api.Success, status
+	}
+	return api.Failure, status
+}
+
+func (o *operator) computeNodeResource(cluster *models.Cluster, nodeInfo *resourceInfo) (string, error) {
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:75]: Please rename this function to indicate that it is working on multiple nodes.
Perhaps checkResourcesAllNodes or computeResourcesAllNodes

File: internal/operators/ocs/validations.go [hunk 27]
```diff
@@ -2,247 +2,251 @@ package ocs
...
 		}
+		o.log.Info("Validate Requirements status ", status)
+		return false, status
 	}
 
-	return countDisks
-
-}
+	if nodeInfo.numDisks%3 != 0 {
+		status = fmt.Sprint("Total disks on the cluster must be a multiple of 3 of size >= 5 GB >= ", minDiskSize, " GB")
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:163]: Please remove the extra >= 5GB

File: internal/operators/ocs/validations.go [hunk 28]
```diff
@@ -2,247 +2,251 @@ package ocs
...
 	return true, status
 }
 
-func (o *ocsValidator) setStatusInsufficientResources(cpu int64, ram int64, disk int64, hostsWithDisks int64, mode string) string {
-	var TotalCPUs int64
-	var TotalRAMGB int64
-	if mode == "Compact Mode" {
-		TotalCPUs = o.OCSRequiredCompactModeCPUCount
-		TotalRAMGB = o.OCSRequiredCompactModeRAMGB
+func validateClusterRequirements(o *operator, nodeInfo *resourceInfo, mode ocsDeploymentMode) bool {
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:193]: I would rename the function to validateRequirementsForMode

File: internal/operators/ocs/validations.go [hunk 29]
```diff
@@ -2,247 +2,251 @@ package ocs
...
 
-func gbToBytes(gb int64) int64 {
-	return gb * int64(units.GB)
-}
-
-// count all disks of drive type ssd or hdd
-func getValidDiskCount(disks []*models.Disk) int {
-
-	var countDisks int
+// used to validate resource requirements for OCS excluding disk requirements and set a status message
```
Comment (nbalacha) [inline internal/operators/ocs/validations.go:149]: Please change the comment - this function now validates the disk requirements as well.

Comment (nbalacha) [reply internal/operators/ocs/validations.go:149]: Would it make sense to rename this function canOCSBeDeployed ()?

Comment (rewantsoni) [reply internal/operators/ocs/validations.go:149]: Yes that would make sense

**Comment by priyanka19-98** [comment]
looks good. The code looks much cleaner. 
Thanks @nbalacha and @rewantsoni

File: internal/operators/ocs/validations.go [hunk 30]
```diff
@@ -2,247 +2,251 @@ package ocs
...
+	OCSRequiredDiskRAMGB           int64             `envconfig:"OCS_REQUIRED_DISK_RAM_GB" default:"5"`    // each disk requires 5GB ram
+	OCSRequiredHosts               int64             `envconfig:"OCS_MINIMUM_HOST" default:"3"`
+	OCSRequiredCPUCount            int64             `envconfig:"OCS_REQUIRED_CPU_COUNT" default:"24"`              // Standard Mode 8*3
+	OCSRequiredRAMGB               int64             `envconfig:"OCS_REQUIRED_RAM_GB" default:"57"`                 // Standard Mode
+	OCSRequiredCompactModeCPUCount int64             `envconfig:"OCS_REQUIRED_COMAPCT_MODE_CPU_COUNT" default:"30"` // Compact Mode cpu requirements for 3 nodes( 4*3(master)+6*3(OCS) cpus)
+	OCSRequiredCompactModeRAMGB    int64             `envconfig:"OCS_REQUIRED_COMAPCT_MODE_RAM_GB" default:"81"`    //Compact Mode ram requirements (8*3(master)+57(OCS) GB)
+	OCSMasterWorkerHosts           int64             `envconfig:"OCS_REQUIRED_MASTER_WORKER_HOSTS" default:"6"`
+	OCSMinimalDeployment           bool              `envconfig:"OCS_MINIMAL_DEPLOYMENT" default:"false"`
```
Comment (priyanka19-98) [inline internal/operators/ocs/validations.go:25]: Just a small change I was thinking. `OCSMinimalDeployment` should be removed.

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:25]: In the manifest generation, we are using the `OCSDeploymentType` to check for different modes. Hence it is not used anywhere else.

Comment (rewantsoni) [reply internal/operators/ocs/validations.go:25]: we are using that in manifest.go

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go:25]: could you point me to where `OCSMinimalDeployment` is used in manifest.go ?

Comment (rewantsoni) [reply internal/operators/ocs/validations.go]: Sorry I read it wrong, I removed it now

Comment (priyanka19-98) [reply internal/operators/ocs/validations.go]: Thanks!

**Review by gamli75 — APPROVED** [review]
